### ISSUE TITLE
refreshToken Not Working with Gmail

### ISSUE THREAD

**Issue by OmriAharon** [nodemailer/nodemailer#910]
After generating new access token, sending mails work for one hour (duration of access token until expires), but after that it doesn't work. I expected `refreshToken` to handle this.
I'm hosting this code on Lambda, so perhaps this makes a difference somehow.
Transporter setup:

```js
const transporter = nodemailer.createTransport({
         host: 'smtp.gmail.com',
         port: 465,
         secure: true,
         auth: {
           type: 'OAuth2',
           user: 'myemail@gmail.com',
           accessToken: 'myAccessToken',
           refreshToken: 'myRefreshToken'
         }
       });
```

The error I get after the access token expires:
```
{ Error: Invalid status code 400
at ClientRequest.req.on.res (/var/task/node_modules/nodemailer/lib/fetch/index.js:205:23)
at emitOne (events.js:96:13)
at ClientRequest.emit (events.js:188:7)
at HTTPParser.parserOnIncomingClient [as onIncoming] (_http_client.js:473:21)
at HTTPParser.parserOnHeadersComplete (_http_common.js:99:23)
at TLSSocket.socketOnData (_http_client.js:362:20)
at emitOne (events.js:96:13)
at TLSSocket.emit (events.js:188:7)
at readableAddChunk (_stream_readable.js:176:18)
at TLSSocket.Readable.push (_stream_readable.js:134:10)
type: 'FETCH',
sourceUrl: 'https://accounts.google.com/o/oauth2/token',
code: 'EAUTH',
command: 'AUTH XOAUTH2' }
```

All the stackoverflow questions I saw got me to this setup and I didn't find an answer in any of those. Only found this issue which is exactly like mine but there's no answer there - https://stackoverflow.com/questions/45412584/gmail-nodemailer-oath2-refresh-token-not-working

**Comment by andris9** [comment]
Based on that error I’d say that nodemailer tries to request a new access token using the refresh token but google oauth server declines which usually means that the refresh token is somehow not correct

**Comment by OmriAharon** [comment]
@De-Lac I solved it eventually. There was nothing wrong with the token. The wrong thing was that I declared the accessToken/refreshToken in the `nodemailer.createTransport`'s auth object. 
Where you should place it, instead, is in the `transporter.sendMail`'s first parameter which is the `mailOptions`. 
Once I moved it there the issue was gone.
[Docs](https://nodemailer.com/smtp/oauth2/) are very very misleading.

**Comment by De-Lac** [comment]
like this?
```
var auth = {
    type                 :  'oauth2',
    clientId            : GMAIL_OAUTH.client_id,
    clientSecret    : GMAIL_OAUTH.client_secret
};

// transporter configuration
var transporter = nodemailer.createTransport({
    host: 'smtp.gmail.com',
    port: 465,
    secure: true,
    auth: auth,
});

    var mailOptions = 
        {
            from:    'Hub21  <info@hub21.it>',  // sender address
            to:      receiverEmail,             // list of receivers
            subject: subject,                   // Subject line
            html:    email_template,
            auth : {
                user                :  'info@mydomain.it',
                refreshToken : GMAIL_OAUTH.refresh_token,
                accessToken  : GMAIL_OAUTH.access_token
            }                    
        };

transporter.sendMail(mailOptions, function(error, info){}
```


in this way I get the error
```
error:  message=Invalid status code 401, stack=Error: Invalid status code 401
    at ClientRequest.req.on.res (/Volumes/HD Daniele/Sites/hub21/server/node_modules/nodemailer/lib/fetch/index.js:221:23)
    at emitOne (events.js:115:13)
    at ClientRequest.emit (events.js:210:7)
    at HTTPParser.parserOnIncomingClient [as onIncoming] (_http_client.js:564:21)
    at HTTPParser.parserOnHeadersComplete (_http_common.js:116:23)
    at TLSSocket.socketOnData (_http_client.js:453:20)
    at emitOne (events.js:115:13)
    at TLSSocket.emit (events.js:210:7)
    at addChunk (_stream_readable.js:252:12)
    at readableAddChunk (_stream_readable.js:239:11)
    at TLSSocket.Readable.push (_stream_readable.js:197:10)
    at TLSWrap.onread (net.js:589:20), type=FETCH, sourceUrl=https://accounts.google.com/o/oauth2/token, code=EAUTH, command=AUTH XOAUTH2
2018/09/04 12:50 - error:  message=Invalid status code 401, stack=Error: Invalid status code 401
    at ClientRequest.req.on.res (/Volumes/HD Daniele/Sites/hub21/server/node_modules/nodemailer/lib/fetch/index.js:221:23)
    at emitOne (events.js:115:13)
    at ClientRequest.emit (events.js:210:7)
    at HTTPParser.parserOnIncomingClient [as onIncoming] (_http_client.js:564:21)
    at HTTPParser.parserOnHeadersComplete (_http_common.js:116:23)
    at TLSSocket.socketOnData (_http_client.js:453:20)
    at emitOne (events.js:115:13)
    at TLSSocket.emit (events.js:210:7)
    at addChunk (_stream_readable.js:252:12)
    at readableAddChunk (_stream_readable.js:239:11)
    at TLSSocket.Readable.push (_stream_readable.js:197:10)
    at TLSWrap.onread (net.js:589:20), type=FETCH, sourceUrl=https://accounts.google.com/o/oauth2/token, code=EAUTH, command=AUTH XOAUTH2


```

**Comment by De-Lac** [comment]
if I try mailOptions without the subobj 'auth':
```
    var mailOptions = 
        {
            from:    '###',
            to:      '###',
            subject:'###',
            html:    '###',
            
            user         : "info@mydomain.it",
            refreshToken : GMAIL_OAUTH.server.refresh_token,
            accessToken  : GMAIL_OAUTH.server.access_token           
        };
```

I get this error

```
 error:  message=Can't create new access token for user, stack=Error: Can't create new access token for user
    at XOAuth2.generateToken (/Volumes/HD Daniele/Sites/hub21/server/node_modules/nodemailer/lib/xoauth2/index.js:172:33)
    at XOAuth2.getToken (/Volumes/HD Daniele/Sites/hub21/server/node_modules/nodemailer/lib/xoauth2/index.js:123:18)
    at SMTPConnection._handleXOauth2Token (/Volumes/HD Daniele/Sites/hub21/server/node_modules/nodemailer/lib/smtp-connection/index.js:1509:27)
    at SMTPConnection.login (/Volumes/HD Daniele/Sites/hub21/server/node_modules/nodemailer/lib/smtp-connection/index.js:356:22)
    at connection.connect (/Volumes/HD Daniele/Sites/hub21/server/node_modules/nodemailer/lib/smtp-transport/index.js:270:32)
    at SMTPConnection.once (/Volumes/HD Daniele/Sites/hub21/server/node_modules/nodemailer/lib/smtp-connection/index.js:188:17)
    at Object.onceWrapper (events.js:314:30)
    at emitNone (events.js:105:13)
    at SMTPConnection.emit (events.js:207:7)
    at SMTPConnection._actionEHLO (/Volumes/HD Daniele/Sites/hub21/server/node_modules/nodemailer/lib/smtp-connection/index.js:1128:14)
    at SMTPConnection._processResponse (/Volumes/HD Daniele/Sites/hub21/server/node_modules/nodemailer/lib/smtp-connection/index.js:762:20)
    at SMTPConnection._onData (/Volumes/HD Daniele/Sites/hub21/server/node_modules/nodemailer/lib/smtp-connection/index.js:558:14)
    at TLSSocket._socket.on.chunk (/Volumes/HD Daniele/Sites/hub21/server/node_modules/nodemailer/lib/smtp-connection/index.js:510:47)
    at emitOne (events.js:115:13)
    at TLSSocket.emit (events.js:210:7)
    at addChunk (_stream_readable.js:252:12), code=EAUTH, command=AUTH XOAUTH2
2018/09/04 12:59 - error:  message=Can't create new access token for user, stack=Error: Can't create new access token for user
    at XOAuth2.generateToken (/Volumes/HD Daniele/Sites/hub21/server/node_modules/nodemailer/lib/xoauth2/index.js:172:33)
    at XOAuth2.getToken (/Volumes/HD Daniele/Sites/hub21/server/node_modules/nodemailer/lib/xoauth2/index.js:123:18)
    at SMTPConnection._handleXOauth2Token (/Volumes/HD Daniele/Sites/hub21/server/node_modules/nodemailer/lib/smtp-connection/index.js:1509:27)
    at SMTPConnection.login (/Volumes/HD Daniele/Sites/hub21/server/node_modules/nodemailer/lib/smtp-connection/index.js:356:22)
    at connection.connect (/Volumes/HD Daniele/Sites/hub21/server/node_modules/nodemailer/lib/smtp-transport/index.js:270:32)
    at SMTPConnection.once (/Volumes/HD Daniele/Sites/hub21/server/node_modules/nodemailer/lib/smtp-connection/index.js:188:17)
    at Object.onceWrapper (events.js:314:30)
    at emitNone (events.js:105:13)
    at SMTPConnection.emit (events.js:207:7)
    at SMTPConnection._actionEHLO (/Volumes/HD Daniele/Sites/hub21/server/node_modules/nodemailer/lib/smtp-connection/index.js:1128:14)
    at SMTPConnection._processResponse (/Volumes/HD Daniele/Sites/hub21/server/node_modules/nodemailer/lib/smtp-connection/index.js:762:20)
    at SMTPConnection._onData (/Volumes/HD Daniele/Sites/hub21/server/node_modules/nodemailer/lib/smtp-connection/index.js:558:14)
    at TLSSocket._socket.on.chunk (/Volumes/HD Daniele/Sites/hub21/server/node_modules/nodemailer/lib/smtp-connection/index.js:510:47)
    at emitOne (events.js:115:13)
    at TLSSocket.emit (events.js:210:7)
    at addChunk (_stream_readable.js:252:12), code=EAUTH, command=AUTH XOAUTH2
```

**Comment by OmriAharon** [comment]
The first option is how I use it. I also pass in the `expires` property with the value you receive when you got your tokens. Looks to me like maybe you used tokens which are invalidated?
Try again with new tokens.

**Comment by De-Lac** [comment]
How can you know the exact value to put in 'expires'? I get the tokens from the playground

**Comment by De-Lac** [comment]
By the way, with the field `expires`, it works! I just put the current epoch time. The provided `access token` results always invalid, so it asks a new token each time (and it works now).

**Comment by Robertvera** [comment]
Thanks god I reached this thread. Pretty misleading documentation indeed, let's see if Nodemailer guys can take a look at it.

**Comment by dkaushikl** [comment]
@De-Lac what is wrong with this


`const nodemailer = require('nodemailer');
const Email = require("email-templates");

const refreshToken = 'refreshToken';
const accessToken = 'accessToken';

const transporter = nodemailer.createTransport({
  service: "Gmail",
  auth: {
    type: 'OAuth2',
    user: 'mydomnain@domain.io',
    clientId: 'clientid.apps.googleusercontent.com',
    clientSecret: 'clientsecret',
  }
});

const email = new Email({
  message: { from: "mydomnain@domain.io", },
  preview: false,
  transport: { jsonTransport: false },
});

resetPasswordMailToManager: function (req, name, key, mailId) {
    const url = `${req.protocol}://${req.get('host')}`;
    email.send({
      template: "resetPassword",
      message: {
        subject: 'Reset your password',
        to: mailId,
        auth: {
          user: 'mydomnain@domain.io',
          refreshToken: refreshToken,
          accessToken: accessToken,
          expires: 1484314697598
        }
      },
      locals: {
        name: name,
        resetPasswordLink: `${url}:/reset-password/${key}`
      }
    }).then(result => {
      transporter.sendMail(result.originalMessage, (error, data) => {
        if (error) {
          console.log(`error`, error);
        }
      });
    })
  }`

![image](https://user-images.githubusercontent.com/13487882/52803814-d2c0cb00-30a8-11e9-80bc-40a9baa25800.png)

**Comment by ismail-raza** [comment]
I'm also have the exact issue. I've used multiple functions for email sending using Nodemailer, but only one of them generating this error on live server. same function is working fine on localhost. does anyone have idea? Please help if anyone resolved this issue till now!

**Comment by beniwalamit123** [comment]
I know i am too late on this post. But i came here as i was facing same issue yesterday. I have fixed the issue and the issue was a silly mistake which was not mentioned anywhere.

So issue was in the process to create refresh token and access token in the google playground page. I selected correct scope (https://mail.google.com) but while creating code and refresh token i did not added my app client id and secret and google was taking a client id and secret of playground by default. So be careful to add your app's client id and secret in the playground.

To add your client id and password. Click on right top corner setting button. Click on checkbox saying use your own credentials and click on close. now if we create access refresh token Google will ask permission for our app instead of Google Playground.

It will resolve the 401 issue. This was not mentioned anywhere on any card created.
